### DISCUSSION TITLE
How to get `gix::objs::Tree` from `repo.head_commit()`?

### DISCUSSION THREAD

**Discussion by Reverier-Xu** [GitoxideLabs/gitoxide#795] [Q&A] [answered]
Hi, thank you for your awesome work!

I'm trying to write a git server with gitoxide, and I'd like to allow user upload a file in web page.

I learned your [example here](https://github.com/Byron/gitoxide/blob/main/gix/examples/init-repo-and-commit.rs) and wrote code like this:

```rust
use gix::{
    objs::{tree, Tree},
    ThreadSafeRepository,
};

...

let mut repo = self.repo.to_thread_local();
let blob_id = repo.write_blob(data)?.into();
let entry = tree::Entry {
    mode: tree::EntryMode::Blob,
    oid: blob_id,
    filename: filename.as_ref().to_owned().into(),
};

let tree = repo.head_commit()?.tree()?;  // <- this `tree` is gix::Tree, which is not same as gix::objs::Tree.

//tree.entries.push(entry);
//let new_tree_id = repo.write_object(tree);
//repo.commit(...)
```

I'm trying to find something to convert two different `Tree`s, there is also no way to construct a `gix::objs::Tree` with old trees in previous commits.

What should I do to implement this? Thanks for your help!

**Comment by Byron** [comment, marked as answer]
Thanks for making me aware. Creating trees for commits indeed is a very manual process right now, and would you are trying to do should definitely be supported.

That said, in `main` you will now find [7c2e5c8d08e4dd1ec115ae06f20f9c8f93d6d616](https://github.com/Byron/gitoxide/commit/7c2e5c8d08e4dd1ec115ae06f20f9c8f93d6d616) which makes it possible to turn a `gix::Tree` into a `gix::objs::Tree` with `tree.try_into()`.

I hope that helps.
